### ISSUE TITLE
Update ESMify to generate ONAM subrecords

### ISSUE THREAD

**Issue by lojack5** [wrye-bash/wrye-bash#29]
[sourceforge](https://sourceforge.net/p/oblivionworks/enhancements/247/)

For Skyrim (and for the two Fallouts) it isn't enough anymore to simply flip the ESM flag on when converting a file to an ESM. Whether this is through the "copy to ESM" function or via the "ESMify Self" function, Bash should generate the needed ONAM subrecords for the TES4 Header so that the file will be fully functional if the user is doing this to play the game with.
Without doing this, a flag-flipped file will have any references it places in vanilla cells completely ignored.
Opposite of this, when flipping back to an ESP, the ONAM subrecords should be stripped from the file as they are not processed by ESP files at all.
Not a major priority at the moment since TES5Edit can do this now, but it would still be nice to have the support be complete.

**Comment by Utumno** [comment]
Thanks @Infernio - really appreciated as always :)

Could you finalize the bsa branch so I can merge? I mean if there needs be some squashing, or some readme edits on the menu - drop the links you used in there, as I saw in discord it's not the simplest topic in the world

**Comment by Utumno** [comment]
That's also an excellent opportunity to start having a look at the records code @Infernio - revisiting the records and the patcher is probably the goal for 308 - complete with tests and docs

**Comment by CptMcSplody** [comment]
@Infernio Just dangling my feet so far by testing with xEdit ONAM updater, but work away on it if you feel like doing it.

@Utumno record updates are on https://github.com/wrye-bash/wrye-bash/commits/shar-mcs-record-patcher-updates the build I threw up on discord was to see if anything broke so miht be missing a few changes. Not sure what these were for https://github.com/wrye-bash/wrye-bash/commit/7c56789d2d5bffadb130366fe9a831578511464f

**Comment by Infernio** [comment]
My current WIP list for 'what needs ONAM' for all games is:

|    Game    |                    Records                     |                                 Source                                  |
|------------|------------------------------------------------|-------------------------------------------------------------------------|
| Oblivion   | N/A                                            | [UESP](https://en.uesp.net/wiki/Tes4Mod:Mod_File_Format/TES4)           |
| Fallout 3  | ACHR, ACRE, LAND, NAVM, PBEA, PGRE, PMIS, REFR | [fopdoc](https://tes5edit.github.io/fopdoc/Fallout3/Records/TES4.html)  |
| Fallout NV | ACHR, ACRE, LAND, NAVM, PMIS, PBEA, PGRE, REFR | [fopdoc](https://tes5edit.github.io/fopdoc/FalloutNV/Records/TES4.html) |
| Skyrim     | ACHR, LAND, NAVM, PGRE, PHZD, REFR             | [UESP](https://en.uesp.net/wiki/Tes5Mod:Mod_File_Format/TES4)           |
| Fallout 4  | Same as Skyrim?                                | Haven't found one yet                                                   |

Not sure what to do about FO4, I might just ask Elminster if he can give me a list for each game.

**Comment by Infernio** [comment]
Bad news, this is going to have to wait a while:
1. https://github.com/wrye-bash/wrye-bash/commit/9072238c20b557ddf34bf6f765f4e50dd7e51887#commitcomment-32015096
   - The way to fix this is to refactor decent parts of `parsers.LoadFactory` and related classes to use game-specific constants.
   - For example, see `LoadFactory.getCellTypeClass`:
   ```python
   def getCellTypeClass(self):
    """Returns type_class dictionary for cell objects."""
    if not self.cellType_class:
        types = ('REFR','ACHR','ACRE','PGRD','LAND','CELL','ROAD')
        getterRecClass = self.getRecClass
        self.cellType_class.update((x,getterRecClass(x)) for x in types)
    return self.cellType_class
   ```
   - This is completely Oblivion-specific - `PGRD` and `ACRE` don't even exist in Skyrim, for example.
2. I don't have a clean way to detect whether or not something is a master, which is required for ONAM detection.
   - To keep it short: In FO3, FNV, Skyrim and Skyrim SE, a plugin needs ONAM data if it is a master. In FO4, all plugins need ONAM data.
   - However, there is no clean 'is_master' method - `ModInfo.isEsm` is the closest, but that one doesn't work correctly on SSE and FO4.
   - In fact, this leads to a direct bug in Wrye Bash - try downloading [this file](https://github.com/wrye-bash/wrye-bash/files/2780927/NotAMaster.esm.zip) and dropping it into your SSE Data folder - Wrye Bash will not detect it as a master. This behavior is correct in previous games, but SSE and FO4 break it - in those games, that file should be a master.

**Comment by Infernio** [comment]
The solution to the second point is either a hack, or a big refactoring of `FileInfo`, `ModInfo`, and all related classes - and I think I'm going to aim at the second option:

1. This will require the `wip-eslsupport-imports` branch to be finalized and merged first, in order to avoid conflicts.
2. `ModInfo`, `FileInfo`, etc. will be moved from `bosh/__init__.py` into a new file - say, `bosh/file_api.py`.
3. Refactoring to turn `file_api.py` into a clean API, inspired by `bsa_files.py` and `save_headers.py`, with tests and everything.
4. Now, game-specific changes can easily be introduced - for example, my `needs_onam` method could be very easily placed into `ModInfo` and overriden in `SkyrimSeModInfo`.

**Comment by Utumno** [comment]
> but that one doesn't work correctly on SSE and FO4.

laod_order.is_master_block ?

> In fact, this leads to a direct bug in Wrye Bash - try downloading this file and dropping it into your SSE Data folder

You mean it's not painted blue? That's an easy fix in esl branch let me do

----

> I think I'm going to aim at the second option:

Seconded! _But_ because I want you around, please don't aim too high (or rather wide). Lots of people after some coding in Bash decided to launch a rewrite and all of them invariably got tired and left half baked code behind. The only exception is me that I actually did rewrite almost all of it - except the patcher/wizard

So actually I am really thrilled by your interest

> This will require the wip-eslsupport-imports branch to be finalized and merged first, in order to avoid conflicts.

Definitively and please do this for me - Now that you are familiar with records loading code, the spec is half done in #429 - so if undecodable, reason ="use xedit", else check high forms (plus the rest of the code)

A commit over my branch would suffice and let me finalize the merge

> ModInfo, FileInfo, etc. will be moved from bosh/__init__.py into a new file - say, bosh/file_api.py.

Hmmmmmmmm - leave them there for now - there are reasons that bosh is still that big, had tried moving stuff out but results were not satisfactory. Hey the BsaFile you did - was the only obvious candidate for dropping (another reason I asked for that)
We 'll get back to that of course!

> Refactoring to turn file_api.py into a clean API, inspired by bsa_files.py and save_headers.py, with tests and everything.

That's what I say don't aim too wide - tests yes sure

> Now, game-specific changes can easily be introduced - for example, my needs_onam method could be very easily placed into ModInfo and overriden in SkyrimSeModInfo.

The current approach is to define such constants into game package (which thanks to @GandaG fetaures now an hierarchy of classes).

My suggestion would be:

1. esl
2. refactor _in place_ to make them game generic/introduce some testing for records code
3. while doing this finalize game support (#358) - investigate how we could move all this to different classes as you discuss. the idea is very nice but the priority is to have working code and it's much easier to refactor existing (read _feasible_) and in the process see how it can be refactored

You do understand I would love a ModInfos hierarchy - bear in  mind I am actually investiagting all this btw #336

**Comment by Infernio** [comment]
Thanks for the link to #336, I hadn't seen that one yet.

Regarding the rest:
Sure, let's place the complete refactoring (with moving things out of `bosh/__init__.py`) on the backburner for now (I wasn't talking about doing that *right now*, it's just what we need to eventually do to fix all these strange game-specific plugin issues).

In the meantime, we should take a look at https://github.com/Wrye-Code-Collection/wrye-bash/commit/b2c83e57abad250fab93b0105194f88f5c750e90 on Sharlikran's repo - that commit implements `ModFile.isEsm` correctly, which will be sufficient for the ONAM stuff for now.
If that could be cherry-picked into the `wip-eslsupport-imports` branch here, then I could continue with ONAM support once that branch is merged.

**Comment by Utumno** [comment]
Haha - https://github.com/wrye-bash/wrye-bash/milestone/13 - https://github.com/wrye-bash/wrye-bash/issues?utf8=%E2%9C%93&q=+is%3Aissue+label%3Abackburner+milestone%3A%22Code+Quality%22

It does not implement isEsm correctly @Infernio - or rather not using it correctly. We need to differentiate between load order and flag handling. `load_order.in_master_block` is the equivalent in my branch - cause I thought that was needed in load order only. Use that one.

Of course I may be wrong and did not draw the line correctly - my esl branch (based on work by Shralikran) is still a WIP - please feel free to correct stuff :)

The wip esl support just has the work by Sharlikran on records - but that needs some polishing. Will rebase on my branch but you need to use my branch (finalizing the esl part there)

**Comment by Infernio** [comment]
I'll take a closer look the entire `wip-eslsupport-imports` branch later today.

**Comment by Utumno** [comment]
Thanks @Infernio - I pushed https://github.com/wrye-bash/wrye-bash/tree/wip-eslsupport-imports-new rebased on  my branch - have a look at that one

**Comment by Infernio** [comment]
I'm moving this one back to 308, there's no way this can be implemented in an efficient way with the current records code. A lot of refactoring and a deep understanding of WB's record code will be necessary for this.

**Comment by Infernio** [comment]
> A lot of refactoring and a deep understanding of WB's record code will be necessary for this.

Now that the former is merged and I have the latter, moving this back to 307 - I have an idea for this, but mostly just doing it so I remember later :)
